codec-http2/src/test/java/io/netty/handler/codec/http2/PriorityStreamWriteDistributorTest.java

```diff
@@ -0,0 +1,663 @@
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+import org.mockito.verification.VerificationMode;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Tests for {@link PriorityStreamWriteDistributor}.
+ */
+public class PriorityStreamWriteDistributorTest {
+    private static final int STREAM_A = 1;
+    private static final int STREAM_B = 3;
+    private static final int STREAM_C = 5;
+    private static final int STREAM_D = 7;
+    private static final int STREAM_E = 9;
+
+    private Http2Connection connection;
+    private PriorityStreamWriteDistributor distributor;
+
+    @Mock
+    private StreamWriteDistributor.Writer writer;
+
+    @Before
+    public void setup() throws Http2Exception {
+        MockitoAnnotations.initMocks(this);
+
+        connection = new DefaultHttp2Connection(false);
+        distributor = new PriorityStreamWriteDistributor(connection);
+
+        // Assume we always write all the allocated bytes.
+        doAnswer(new Answer<Void>() {
+            @Override
+            public Void answer(InvocationOnMock in) throws Throwable {
+                Http2Stream stream = (Http2Stream) in.getArguments()[0];
+                int numBytes = (Integer) in.getArguments()[1];
+                int streamableBytes = distributor.unallocatedStreamableBytes(stream) - numBytes;
+                updateStream(stream.id(), streamableBytes, streamableBytes > 0);
+                return null;
+            }
+        }).when(writer).write(any(Http2Stream.class), anyInt());
+
+        connection.local().createStream(STREAM_A, false);
+        connection.local().createStream(STREAM_B, false);
+        Http2Stream streamC = connection.local().createStream(STREAM_C, false);
+        Http2Stream streamD = connection.local().createStream(STREAM_D, false);
+        streamC.setPriority(STREAM_A, DEFAULT_PRIORITY_WEIGHT, false);
+        streamD.setPriority(STREAM_A, DEFAULT_PRIORITY_WEIGHT, false);
+    }
+
+    @Test
+    public void bytesUnassignedAfterProcessing() {
+        updateStream(STREAM_A, 1, true);
+        updateStream(STREAM_B, 2, true);
+        updateStream(STREAM_C, 3, true);
+        updateStream(STREAM_D, 4, true);
+
+        assertFalse(write(10));
+        verifyWrite(STREAM_A, 1);
+        verifyWrite(STREAM_B, 2);
+        verifyWrite(STREAM_C, 3);
+        verifyWrite(STREAM_D, 4);
+
+        assertFalse(write(10));
+        verifyWrite(STREAM_A, 0);
+        verifyWrite(STREAM_B, 0);
+        verifyWrite(STREAM_C, 0);
+        verifyWrite(STREAM_D, 0);
+    }
+
+    @Test
+    public void bytesUnassignedAfterProcessingWithException() {
+        updateStream(STREAM_A, 1, true);
+        updateStream(STREAM_B, 2, true);
+        updateStream(STREAM_C, 3, true);
+        updateStream(STREAM_D, 4, true);
+
+        Exception fakeException = new RuntimeException("Fake exception");
```

Comment:
Be nice if we have a utility class for this.  See https://github.com/netty/netty/issues/4274